pkg/addonmanager/controllers/agentdeploy/controller.go

```diff
@@ -150,20 +151,14 @@ func (c *addonDeployController) sync(ctx context.Context, syncCtx factory.SyncCo
 	}
 
 	cluster, err := c.managedClusterLister.Get(clusterName)
-	if errors.IsNotFound(err) {
-		// to delete the addon to make sure the addon is cleaned up when the cluster is not found.
-		// need go to sync to handle cleanup if the addon is deleting.
-		// in the normal case the addon should have been cleaned up after the cluster is deleted.
-		if addon.DeletionTimestamp.IsZero() {
-			return c.cleanupAddon(ctx, clusterName, addonName)
-		}
-	}
 	if err != nil {
-		return err
+		// TODO: should consider to force delete the addon and its manifestWork if cluster is not found.
+		return client.IgnoreNotFound(err)
```

Comment:
I do not understand, if the cluster is not found, shouldn't we remove the work?

Reply:
the cluster will be nil if get `not found` here, so cannot go into sync with a nil cluster because there are many functions using cluster without checking if cluster is valid or not.
add a TODO here to force delete the addon and manifestWork if the cluster is not found when we begin to handle cluster deletion in the future.

Reply:
but we do remove manifestwork when cluster is not found, shouldn't this change break the previous behavior?

Reply:
we only return nil when cluster is not found  before my refactor PR.  I deleted addon but not manifestWork when cluster is not found in my refactor PR.  
This is an exception scenario, so I think only to delete addon is not a good option, we should force deleted addon and manifestWork in this scenario, should we do the force deletion in addon-framework or other component like deletion controller of registration?